### DISCUSSION TITLE
Организация обмена с 1с

### DISCUSSION THREAD

**Discussion by LostAlly** [totumonline/totum-ru-issues-and-discussions#510] [Как это сделать?] [answered]
Делаю синхронизацию справочников с 1с флеш через odata (односторонняя из 1с в тотум)
Возникло несколько вопросов
1. В 1с для связи используются guid у каждого объекта есть такой глобальный уникальный номер.
Вопрос в том где его хранить - в тех же таблицах где и информация или сделать отдельную таблицу соответствия объектов Тотума и 1с это позволит в случае чего убрать или заменить 1с на что то другое. Или я заморачиваюсь и нужно делать поля этих айдишников в самих справочниках и таблицах.
2. Как контролировать связь объектов между собой id Тотума или использовать guid 1с
3. Обмен будет раздельно односторонним - из 1с в Тотум справочники, из Тотума в 1с часть документов счета, закрывающие, производственные(по поводу склада еще думаю).

Одним словом - как лучше организовать обмен, спасибо.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Че делали мы в реальных проектах:

1. GUID делали строковое поле в таблице Тотума в которой должен лежать Справочник. Обычно первым полем, так и называли guid тк это строка. Для пользователей людей это поле может быть скрыто.
2. Если вы только забираете справочники, то внутри Тотума этот guid как козе пропеллер. Он нужен только для синхронизации с 1c.

**НО!!! если из 1с приходят Документы или вы их отправляете обратно** — то вы обязаны использовать именно GUID в том числе внутри тотума в Селектах через параметр bfield.

3. Вот собственно для этого вам и нужны гиды в тотум, тк когда вы отправите документ обратно и вам нужно будет передать, что он относится к позиции n вам нужно будет отправлять именно гид.

**Reply by ttm-alexeypolunin** [reply]
Соответсвенно при обновлении справочника вы:

1. Отправляете только изменившиеся/добавившиеся позиции + признак удаления отдельно (это хороший вариант, тк справочники могут быть очень большие и не пролезать через кучу всего, да и любое, что обрабатывается дольше минуты всегда опасная штука)
2. На входе по guid в соответсвующем справочнике смотрите есть ли такой (просто берете select и если он непустой, значит есть). Тогда используете id полученный из проверочного сеекта для замены значения. Если пустой то инсертите новую строку.

Если нет возможности отправлять только измененные (хз как там в кишках 1С, можно триггер написать или нет), то при больших справочниках, отправляйте его частями.

**Reply by ttm-alexeypolunin** [reply]
На каждый справочник сделайте свой ремоут — в таблице ремоутов можно добавить им поле с классификатором. И отправляйте справочники по одному, если скопом, то легко можно запутаться и в справочник Ножек для стульев надобавлять Диванных подушек.

**Reply by ttm-alexeypolunin** [reply]
Раздельные ремоуты нивелируют ту проблему, для которой нужна была-бы центральная таблица uid-ов.

**Reply by LostAlly** [reply]
Я не использую ремоты, во фреше механизм обмена инициируется через обращение к базе 1с посредством формирования get строки запросов, документы засыпаются в 1с похоже так же(я еще разбираюсь). Т.е. получается что инициирует обмен (получение и отправку) Тотум по сути дела пообъектно обмениваясь json’ами.

Не думал о вызовах тотума из 1с, нужно глянуть этот момент, подозреваю стандартно 1с не имеет вебхуков.

Поэтому я думал, получать справочники автоматом раз в сутки по крону, просто перезаписывая значения все, так как при Данном обмене 1с не может инициировать связь с тотумом при каких то изменениях у себя.

Это все навело меня на мысль - если Тотум создает документы отсылает в 1с, а потом внести в них изменения в 1с, тотум не узнает про эти изменения, не опрашивать же все документы каждый раз. Нужно крепко переподумать.

 Спасибо.
